sympy/stats/tests/test_stochastic_process.py

```diff
@@ -32,19 +32,39 @@ def test_DiscreteMarkovChain():
 
     raises(ValueError, lambda: sample_stochastic_process(t))
     raises(ValueError, lambda: next(sample_stochastic_process(X)))
+
     # pass name and state_space
-    Y = DiscreteMarkovChain("Y", [1, 2, 3])
-    assert Y.transition_probabilities == None
-    assert Y.state_space == FiniteSet(1, 2, 3)
-    assert P(Eq(Y[2], 1), Eq(Y[0], 2)) == Probability(Eq(Y[2], 1), Eq(Y[0], 2))
-    assert E(X[0]) == Expectation(X[0])
+    sym = symbols('a', real=True)
+    # any hashable object should be a valid state
+    state_spaces = [[1, 2, 3], [-1, pi, None], ['Hello', sym, DiscreteMarkovChain]]
+    chains = [DiscreteMarkovChain("Y", state_spaces[0]),
+              DiscreteMarkovChain("Y", state_spaces[1]),
+              DiscreteMarkovChain("Y", state_spaces[2])]
+    for i, Y in enumerate(chains):
+        assert isinstance(Y.transition_probabilities, MatrixSymbol)
+        assert Y.state_space == Tuple(*state_spaces[i])
+
+        with ignore_warnings(UserWarning):  # TODO: Restore tests once warnings are removed
+            assert P(Eq(Y[2], 1), Eq(Y[0], 2), evaluate=False) == Probability(Eq(Y[2], 1), Eq(Y[0], 2))
+        assert E(Y[0]) == Expectation(Y[0])
+
+        raises(ValueError, lambda: next(sample_stochastic_process(Y)))
+
     raises(TypeError, lambda: DiscreteMarkovChain("Y", dict((1, 1))))
-    raises(ValueError, lambda: next(sample_stochastic_process(Y)))
+
+    # pass name and transition_probabilities
+    chains = [DiscreteMarkovChain("Y", trans_probs=Matrix([[]])),
+              DiscreteMarkovChain("Y", trans_probs=Matrix([[0, 1], [1, 0]])),
+              DiscreteMarkovChain("Y", trans_probs=Matrix([[pi, 1-pi], [sym, 1-sym]]))]
+    for Z in chains:
+        assert Z.num_states == Z.transition_probabilities.shape[0]
+        assert isinstance(Z.transition_probabilities, ImmutableDenseMatrix)
+        assert isinstance(Z.state_space, Tuple)
 
     # pass name, state_space and transition_probabilities
     T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
     TS = MatrixSymbol('T', 3, 3)
-    Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
```

Comment:
Is this test still working with the new patch?

Reply:
Yes. There's just lots of tests with the default kind of range(n) state space so I decided to change it to something that would have broken the old version. I still need to add a few more tests I think.